Make mmap-ing of value log optional.

deepakjois opened the pull request:
We introduce a new option ValueLogLoadingMode, which controls
whether the value log is read via mmap, or by standard File I/O.

Methods that fall in the path to reading the value from log now
take an additional argument for the slice which would be used to
read the data, if the ValueLoadLoadingMode is set to FileIO.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/badger/341)
<!-- Reviewable:end -->

manishrjain commented:
@janardhan1993 Can you have a look as well? This is a deep change and can easily have bugs. So, we need to be thorough with this one.

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[value.go, line 143 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/341#-L-zTtOP91wnxKH4nOR9:-L-zTtOP91wnxKH4nORA:bamomk3) ([raw file](https://github.com/dgraph-io/badger/blob/8ba769be7ad9692bac1654b4ec91bddc06617374/value.go#L143)):*
> ```Go
> 
> func (lf *logFile) doneWriting(offset uint32) error {
> 	// Sync before acquiring lock.  (We call this from writvlog.Readvlog.Reade() and thus know we have shared access
> ```

Looks like a typo.

---

*[value.go, line 159 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/341#-L-zTzfrgeiCb6fy6VPA:-L-zTzfsIpVADrJjC_6k:b76911o) ([raw file](https://github.com/dgraph-io/badger/blob/8ba769be7ad9692bac1654b4ec91bddc06617374/value.go#L159)):*
> ```Go
> 		return err
> 	}
> 	defer lf.lock.Unlock()
> ```

This change of behavior looks buggy.

---

*[value.go, line 814 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/341#-L-zUtoY7BLwqlS1TaS-:-L-zUtoY7BLwqlS1TaS0:b-cajrix) ([raw file](https://github.com/dgraph-io/badger/blob/8ba769be7ad9692bac1654b4ec91bddc06617374/value.go#L814)):*
> ```Go
> 
> 	buf, err := lf.read(vp, s)
> 	return buf, lf.lock.RUnlock, err
> ```

If the file is not mmaped, we don't need to hold a lock onto it, because we're copying the value directly into slice anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/341#-:-L-zVG-oOq-pWPTSza9P:baj8a2o)*
<!-- Sent from Reviewable.io -->

janardhan1993 commented:
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[iterator.go, line 152 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/341#-L01uG3h0hCbmRdWFihr:-L01uG3h0hCbmRdWFihs:bhau0bj) ([raw file](https://github.com/dgraph-io/badger/blob/8ba769be7ad9692bac1654b4ec91bddc06617374/iterator.go#L152)):*
> ```Go
> 	if item.db.opt.ValueLogLoadingMode == options.MemoryMap {
> 		buf := item.slice.Resize(len(val))
> 		copy(buf, val)
> ```

We hold a lock over vlog and don't munmap, do we still need to copy here ?

---

*[value.go, line 125 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/341#-L01ucMR92OSDlBAJUK-:-L01ucMR92OSDlBAJUK0:bc1yt85) ([raw file](https://github.com/dgraph-io/badger/blob/8ba769be7ad9692bac1654b4ec91bddc06617374/value.go#L125)):*
> ```Go
> 		buf = s.Resize(int(p.Len))
> 		var n int
> 		n, err = lf.fd.ReadAt(buf, int64(offset))
> ```

Throw error if n < offset

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/341)*
<!-- Sent from Reviewable.io -->

deepakjois commented:
Review status: 5 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[value.go, line 125 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/341#-L01ucMR92OSDlBAJUK-:-L030fjukJf10MU36LV9:bz191y0) ([raw file](https://github.com/dgraph-io/badger/blob/8ba769be7ad9692bac1654b4ec91bddc06617374/value.go#L125)):*
`n` and `offset` are not comparable. Do you mean `n < len(buf)`? If so that is already covered.

> ReadAt always returns a non-nil error when n < len(b).

---

*[value.go, line 143 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/341#-L-zTtOP91wnxKH4nOR9:-L02zCAI-iffNZ5qv82z:b-886gsx) ([raw file](https://github.com/dgraph-io/badger/blob/8ba769be7ad9692bac1654b4ec91bddc06617374/value.go#L143)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Looks like a typo.
</blockquote></details>

Fixed

---

*[value.go, line 159 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/341#-L-zTzfrgeiCb6fy6VPA:-L031UsgjrN7A6duJUtQ:bnkfa0r) ([raw file](https://github.com/dgraph-io/badger/blob/8ba769be7ad9692bac1654b4ec91bddc06617374/value.go#L159)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This change of behavior looks buggy.
</blockquote></details>

I guess it is a typo. I intended to have `defer lf.lock.Unlock()` before doing `lf.munmap()`.

Fixed now.

---

*[value.go, line 814 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/341#-L-zUtoY7BLwqlS1TaS-:-L033W-jdxkjcQ3bIQdg:bh79iux) ([raw file](https://github.com/dgraph-io/badger/blob/8ba769be7ad9692bac1654b4ec91bddc06617374/value.go#L814)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

If the file is not mmaped, we don't need to hold a lock onto it, because we're copying the value directly into slice anyway.
</blockquote></details>

Good point.

I have now changed the behavior to immediately unlock if we are using File I/O.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/341)*
<!-- Sent from Reviewable.io -->

deepakjois commented:
Review status: 5 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[iterator.go, line 152 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/341#-L01uG3h0hCbmRdWFihr:-L036GOnIjcpfTdw0dWz:b-9ot0oh) ([raw file](https://github.com/dgraph-io/badger/blob/8ba769be7ad9692bac1654b4ec91bddc06617374/iterator.go#L152)):*
<details><summary><i>Previously, janardhan1993 (Janardhan Reddy) wrote…</i></summary><blockquote>

We hold a lock over vlog and don't munmap, do we still need to copy here ?
</blockquote></details>

I made a small change elsewhere based on Manish’s comment. We dont hold a lock anymore, if ValueLogLoadingMode is File I/O.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/341)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Couple of comments.

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[value.go, line 820 at r2](https://reviewable.io:443/reviews/dgraph-io/badger/341#-L07hD7DWkWTQhHXmcDU:-L07hD7DWkWTQhHXmcDV:b-rxuluq) ([raw file](https://github.com/dgraph-io/badger/blob/1b12c379a429b39cf9a41dcf256ed31327f587f3/value.go#L820)):*
> ```Go
> 	// and return an empty function as callback.
> 	lf.lock.RUnlock()
> 	return buf, func() {}, err
> ```

Can return nil instead of empty func. runCallback checks for nil, before invoking the func.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/341#-:-L07halOkzPgxS4OmqL-:bojlezm)*
<!-- Sent from Reviewable.io -->

deepakjois commented:
Review status: 7 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[value.go, line 820 at r2](https://reviewable.io:443/reviews/dgraph-io/badger/341#-L07hD7DWkWTQhHXmcDU:-L07z6SjiHfw8_nSYHt3:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/1b12c379a429b39cf9a41dcf256ed31327f587f3/value.go#L820)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Can return nil instead of empty func. runCallback checks for nil, before invoking the func.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/341)*
<!-- Sent from Reviewable.io -->
